### PR TITLE
python3Packages.nanobind: 2.0.0 -> 2.1.0

### PR DESCRIPTION
**Pull request by phiadaarr**
## Description of changes

Update to latest version and change dependency from `tensorflow` to `tensorflow-bin` as requested in #336897

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/nanobind/default.nix [hunk 1]
```diff
@@ -1,31 +1,31 @@
...
+  pathspec,
+  scikit-build-core,
   scipy,
+  tensorflow-bin,
   torch,
-  jax,
-  jaxlib,
-  tensorflow,
-  setuptools,
 }:
```
Comment (GaetanLepage) [inline pkgs/development/python-modules/nanobind/default.nix:18]: The inputs should be sorted by order of apparition, not alphabetically.
[Example from the documentation](https://nixos.org/manual/nixpkgs/stable/#buildpythonpackage-function)

Comment (phiadaarr) [reply pkgs/development/python-modules/nanobind/default.nix]: done.

**Review by GaetanLepage — APPROVED** [review]
LGTM thanks !

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 337405` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 337405` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.islpy</li>
    <li>python312Packages.islpy.dist</li>
    <li>python312Packages.loopy</li>
    <li>python312Packages.loopy.dist</li>
    <li>python312Packages.nanobind</li>
    <li>python312Packages.nanobind.dist</li>
    <li>python312Packages.pyopencl</li>
    <li>python312Packages.pyopencl.dist</li>
    <li>python312Packages.reikna</li>
    <li>python312Packages.reikna.dist</li>
    <li>python312Packages.sasmodels</li>
    <li>python312Packages.sasmodels.dist</li>
  </ul>
</details>
<details>
  <summary>8 packages failed to build:</summary>
  <ul>
    <li>python311Packages.loopy</li>
    <li>python311Packages.loopy.dist</li>
    <li>python311Packages.pyopencl</li>
    <li>python311Packages.pyopencl.dist</li>
    <li>python311Packages.reikna</li>
    <li>python311Packages.reikna.dist</li>
    <li>python311Packages.sasmodels</li>
    <li>python311Packages.sasmodels.dist</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.islpy</li>
    <li>python311Packages.islpy.dist</li>
    <li>python311Packages.nanobind</li>
    <li>python311Packages.nanobind.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 337405` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>24 packages built:</summary>
  <ul>
    <li>python311Packages.islpy</li>
    <li>python311Packages.islpy.dist</li>
    <li>python311Packages.loopy</li>
    <li>python311Packages.loopy.dist</li>
    <li>python311Packages.nanobind</li>
    <li>python311Packages.nanobind.dist</li>
    <li>python311Packages.pyopencl</li>
    <li>python311Packages.pyopencl.dist</li>
    <li>python311Packages.reikna</li>
    <li>python311Packages.reikna.dist</li>
    <li>python311Packages.sasmodels</li>
    <li>python311Packages.sasmodels.dist</li>
    <li>python312Packages.islpy</li>
    <li>python312Packages.islpy.dist</li>
    <li>python312Packages.loopy</li>
    <li>python312Packages.loopy.dist</li>
    <li>python312Packages.nanobind</li>
    <li>python312Packages.nanobind.dist</li>
    <li>python312Packages.pyopencl</li>
    <li>python312Packages.pyopencl.dist</li>
    <li>python312Packages.reikna</li>
    <li>python312Packages.reikna.dist</li>
    <li>python312Packages.sasmodels</li>
    <li>python312Packages.sasmodels.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 337405` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>24 packages built:</summary>
  <ul>
    <li>python311Packages.islpy</li>
    <li>python311Packages.islpy.dist</li>
    <li>python311Packages.loopy</li>
    <li>python311Packages.loopy.dist</li>
    <li>python311Packages.nanobind</li>
    <li>python311Packages.nanobind.dist</li>
    <li>python311Packages.pyopencl</li>
    <li>python311Packages.pyopencl.dist</li>
    <li>python311Packages.reikna</li>
    <li>python311Packages.reikna.dist</li>
    <li>python311Packages.sasmodels</li>
    <li>python311Packages.sasmodels.dist</li>
    <li>python312Packages.islpy</li>
    <li>python312Packages.islpy.dist</li>
    <li>python312Packages.loopy</li>
    <li>python312Packages.loopy.dist</li>
    <li>python312Packages.nanobind</li>
    <li>python312Packages.nanobind.dist</li>
    <li>python312Packages.pyopencl</li>
    <li>python312Packages.pyopencl.dist</li>
    <li>python312Packages.reikna</li>
    <li>python312Packages.reikna.dist</li>
    <li>python312Packages.sasmodels</li>
    <li>python312Packages.sasmodels.dist</li>
  </ul>
</details>

**Comment by phiadaarr** [comment]
> Result of `nixpkgs-review pr 337405` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
>
> 8 packages failed to build:
> * python311Packages.loopy
> * python311Packages.loopy.dist
> * python311Packages.pyopencl
> * python311Packages.pyopencl.dist
> * python311Packages.reikna
> * python311Packages.reikna.dist
> * python311Packages.sasmodels
> * python311Packages.sasmodels.dist

Do you have any indication what goes wrong on aarch64-darwin? I do not have access to this architecture myself.

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 337405` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.islpy</li>
    <li>python312Packages.islpy.dist</li>
    <li>python312Packages.loopy</li>
    <li>python312Packages.loopy.dist</li>
    <li>python312Packages.nanobind</li>
    <li>python312Packages.nanobind.dist</li>
    <li>python312Packages.pyopencl</li>
    <li>python312Packages.pyopencl.dist</li>
    <li>python312Packages.reikna</li>
    <li>python312Packages.reikna.dist</li>
    <li>python312Packages.sasmodels</li>
    <li>python312Packages.sasmodels.dist</li>
  </ul>
</details>
<details>
  <summary>8 packages failed to build:</summary>
  <ul>
    <li>python311Packages.loopy</li>
    <li>python311Packages.loopy.dist</li>
    <li>python311Packages.pyopencl</li>
    <li>python311Packages.pyopencl.dist</li>
    <li>python311Packages.reikna</li>
    <li>python311Packages.reikna.dist</li>
    <li>python311Packages.sasmodels</li>
    <li>python311Packages.sasmodels.dist</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.islpy</li>
    <li>python311Packages.islpy.dist</li>
    <li>python311Packages.nanobind</li>
    <li>python311Packages.nanobind.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
`pyopencl` will need to be marked as broken on `aarch64-darwin`:

```
[18/18] Linking CXX shared module _cl.cpython-311-darwin.so
FAILED: _cl.cpython-311-darwin.so 
: && /nix/store/gj5jfkr5y5s95vhgvh5hzq70p8dzhw57-clang-wrapper-16.0.6/bin/clang++ -O3 -DNDEBUG  -bundle -Wl,-headerpad_max_install_names -Wl,@/nix/store/71j9ydx7rzpjj6y59bjyv2ha9cqmxxcl-python3.11-nanobind-2.1.0/lib/python3.>
Undefined symbols for architecture arm64:
  "_clCloneKernel", referenced from:
      pyopencl::kernel::clone() in wrap_cl_part_2.cpp.o
  "_clCreateCommandQueueWithProperties", referenced from:
      pyopencl::command_queue::command_queue(pyopencl::context const&, pyopencl::device const*, nanobind::object) in wrap_cl_part_1.cpp.o
  "_clCreatePipe", referenced from:
      pyopencl::create_pipe(pyopencl::pipe*, pyopencl::context const&, unsigned long long, unsigned int, unsigned int, nanobind::sequence) in wrap_cl_part_2.cpp.o
  "_clCreateProgramWithIL", referenced from:
      pyopencl::create_program_with_il(pyopencl::context&, nanobind::bytes const&) in wrap_cl_part_2.cpp.o
  "_clCreateSamplerWithProperties", referenced from:
      pyopencl::sampler::sampler(pyopencl::context const&, nanobind::sequence) in wrap_cl_part_2.cpp.o
      pyopencl::sampler::sampler(pyopencl::context const&, bool, unsigned int, unsigned int) in wrap_cl_part_2.cpp.o
  "_clEnqueueSVMFree", referenced from:
      pyopencl::svm_allocation::release() in wrap_cl_part_2.cpp.o
      pyopencl::svm_allocation::enqueue_release(pyopencl::command_queue*, nanobind::object) in wrap_cl_part_2.cpp.o
      pyopencl::svm_allocator::free(pyopencl::svm_held_pointer&&) in wrap_mempool.cpp.o
  "_clEnqueueSVMMap", referenced from:
      pyopencl::enqueue_svm_map(pyopencl::command_queue&, unsigned int, unsigned long long, pyopencl::svm_pointer&, nanobind::object, nanobind::object) in wrap_cl_part_2.cpp.o
  "_clEnqueueSVMMemFill", referenced from:
      pyopencl::enqueue_svm_memfill(pyopencl::command_queue&, pyopencl::svm_pointer&, nanobind::object, nanobind::object, nanobind::object) in wrap_cl_part_2.cpp.o
  "_clEnqueueSVMMemcpy", referenced from:
      pyopencl::enqueue_svm_memcpy(pyopencl::command_queue&, unsigned int, pyopencl::svm_pointer&, pyopencl::svm_pointer&, nanobind::object, nanobind::object) in wrap_cl_part_2.cpp.o
  "_clEnqueueSVMMigrateMem", referenced from:
      pyopencl::enqueue_svm_migratemem(pyopencl::command_queue&, nanobind::sequence, unsigned long long, nanobind::object) in wrap_cl_part_2.cpp.o
  "_clEnqueueSVMUnmap", referenced from:
      pyopencl::enqueue_svm_unmap(pyopencl::command_queue&, pyopencl::svm_pointer&, nanobind::object) in wrap_cl_part_2.cpp.o
  "_clGetDeviceAndHostTimer", referenced from:
      pyopencl::device::device_and_host_timer() const in wrap_cl_part_1.cpp.o
  "_clGetHostTimer", referenced from:
      pyopencl::device::host_timer() const in wrap_cl_part_1.cpp.o
  "_clGetKernelSubGroupInfo", referenced from:
      pyopencl::kernel::get_sub_group_info(pyopencl::device const&, unsigned int, nanobind::object) in wrap_cl_part_2.cpp.o
  "_clGetPipeInfo", referenced from:
      pyopencl::pipe::get_pipe_info(unsigned int) const in wrap_cl_part_2.cpp.o
  "_clSVMAlloc", referenced from:
      pyopencl::svm_allocation::svm_allocation(nanobind::ref<pyopencl::context> const&, unsigned long, unsigned int, unsigned long long, pyopencl::command_queue const*) in wrap_cl_part_2.cpp.o
      pyopencl::svm_allocator::allocate(unsigned long) in wrap_mempool.cpp.o
  "_clSVMFree", referenced from:
      pyopencl::svm_allocation::release() in wrap_cl_part_2.cpp.o
      pyopencl::svm_allocator::free(pyopencl::svm_held_pointer&&) in wrap_mempool.cpp.o
  "_clSetDefaultDeviceCommandQueue", referenced from:
      pyopencl::context::set_default_device_command_queue(pyopencl::device const&, pyopencl::command_queue const&) in wrap_cl_part_1.cpp.o
  "_clSetKernelArgSVMPointer", referenced from:
      pyopencl::kernel::set_arg_svm(unsigned int, pyopencl::svm_pointer const&) in wrap_cl_part_2.cpp.o
  "_clSetProgramSpecializationConstant", referenced from:
      pyopencl::program::set_specialization_constant(unsigned int, nanobind::object) in wrap_cl_part_2.cpp.o
ld: symbol(s) not found for architecture arm64
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.

*** CMake build failed

**Comment by GaetanLepage** [comment]
I will merge this nonetheless. Can you create a PR (and/or issue) about `pyopencl` please ?

**Comment by GaetanLepage** [comment]
> I will merge this nonetheless. Can you create a PR (and/or issue) about `pyopencl` please ?

Done in https://github.com/NixOS/nixpkgs/pull/337876

**Pull request merged**